Create platform-automation.md (#219)
bcgov/platform-developer-docs: src/docs/platform-architecture-reference/platform-automation.md

---
title: Platform automation

slug: platform-automation

description: Describes the automation we run on to manage the OpenShift platform.

keywords: resource, reclamation, tool, automation, security

page_purpose: Describes the automation we run on to manage the OpenShift platform.

audience: technical lead, developer

author: Matt Spencer, Steven Barre

content_owner: Olena Mitovska

sort_order: 6
---

# Platform Automation

We use some automation on the platform to help manage resource usage, alert about misconfigured objects and to encourage teams to use images that are secure and up to date. 

## Resource reclamation tool

To date, resource utilization (the use of CPU and RAM) on our platforms has not been efficient. For example, almost 50% of CPU capacity on the Silver cluster is reserved by apps, but not actually used. This unused utilization increases costs and impacts our ability to onboard new teams.  
 

> To address this, we’ve developed an automated tool that monitors apps on the platform, looking for inactive or broken deployments. 

Author: Thanks, I've made some changes aimed to do this

### Targets of this tool
 
The tool only monitors `deployments` and `deploymentconfigs` in non-production namespaces (dev, test and tools) and there is no impact to production deployments (prod). This tool does not target databases in `statefulsets`.

The tool looks for two types of deployments: 
- Deployments that are crashing consistently 
- Deployments that have not changed in over a year

> 

Reviewer: ```suggestion

### How it works

Emails are sent to the Product Owner and Technical Leads registered for the Product in the [Registry](https://registry.developer.gov.bc.ca/). If you want to change who gets the emails, go update the contacts in the Registry.

The first tool checks the `lastUpdateTime` of the `Progressing` block in the `status` field of the Deployment or DeploymentConfig. If it is close to a year old, but not over, a warning email will be sent. If the deployment is managed by ArgoCD then a warning email is sent asking for action to be taken. This is because ArgoCD will just scale it back up after the tool scaled it down. If not managed by ArgoCD and the timestamp is over a year then the deployment is scaled to zero replicas and an email is sent.

You can check the timestamp on your deployment in the YAML under `status`.
 
 ```yaml
   conditions:
  - lastTransitionTime: "2021-11-25T17:04:01Z"
    lastUpdateTime: "2021-11-25T17:04:11Z"
    message: replication controller "backup-postgresql-1" successfully rolled out
    reason: NewReplicationControllerAvailable
    status: "True"
    type: Progressing
```

The second tool looks at the restart count of pods. If a pod has more than 100 restarts it looks for the controlling object of the pod, be it Deployment, DeploymentConfig, or StatefulSet. If the object is managed by ArgoCD then a warning email is sent asking for action to be taken. This is because ArgoCD will just scale it back up after the tool scaled it down. If not managed by ArgoCD, the object is scaled to zero replicas and an email is sent.
```

### Responding when deployments are scaled down by this tool

Before you simply scale your pods back up, you should consider fixing the underlying issues that placed your application in one of these two groups in the first place.

>  If your application is crashing regularly, your technical team should investigate and fix the cause of these crashes. 

Reviewer: ```suggestion
 If your application is crashing regularly, your technical team should investigate and fix the cause of these crashes. You can see the number of times your pod has restarted in the web console or via `oc get pods` to get an idea of if it is crashing a lot.
```

>  If your application has not changed in over a year, your technical team should build a new image, which ensures that your application is able to take advantage of any security patches from the past year. 

Reviewer: ```suggestion
 If your application has not changed in over a year, your technical team should build a new image, which ensures that your application is able to take advantage of any security patches from the past year. It is highly likely your image contains Critical security vulnerabilities if it has not been updated in over a year.
```

>  Maintaining your application is part of the [Memorandum of Understanding](https://digital.gov.bc.ca/cloud/services/private/onboard/#memorandum) that teams working on the platform agree to. 

Reviewer: ```suggestion
 Maintaining your application is part of the [Memorandum of Understanding](https://digital.gov.bc.ca/cloud/services/private/onboard/#memorandum) that teams working on the platform agree to. More detailed docs on [maintaining an image](https://docs.developer.gov.bc.ca/maintain-an-application/#maintain-images) are also available.
```

Historically, application teams working on our platform have not efficiently utilized resources such as CPU and RAM. For instance, apps reserve almost 50% of CPU capacity on the Silver cluster without actually using it. This unused utilization increases costs and impacts our ability to onboard new teams. Additionally, some teams neglect to maintain and update their deployment images, leading to security vulnerabilities, incompatibilities, and other problems.

To address this, we’ve developed automated tools that monitor apps on the platform. 
 
These tools looks for these types of objects: 
- `Deployments` and `deploymentconfigs` that are based on images have not changed in over a year
- `Deployments`, `deploymentconfigs` and `statefulsets` that are crashing constantly 

Emails are sent to the Product Owner and Technical Leads registered for the Product in the [Product Registry](https://registry.developer.gov.bc.ca/). To modify the recipients of the emails, please update the contact information in the Registry.

## Deployments and deploymentconfigs that have not changed in a year 

> The tool monitors the `lastUpdateTime` of the `Progressing` block within the `status` field of the Deployment or DeploymentConfig. If this timestamp is approaching one year but has not yet exceeded it, a warning email is triggered. In the case of deployments managed by ArgoCD, an email requests action since ArgoCD may scale it back up after the tool has scaled it down. For deployments not managed by ArgoCD and with a timestamp surpassing one year, the tool scales down the replicas to zero and sends an email notification.

Author: What's the context here @ShellyXueHan, had teams been confused about that? Could you please add a suggestion with what you'd like to see added?

> The tool monitors the `lastUpdateTime` of the `Progressing` block within the `status` field of the Deployment or DeploymentConfig. If this timestamp is approaching one year but has not yet exceeded it, a warning email is triggered. In the case of deployments managed by ArgoCD, an email requests action since ArgoCD may scale it back up after the tool has scaled it down. For deployments not managed by ArgoCD and with a timestamp surpassing one year, the tool scales down the replicas to zero and sends an email notification.

Reviewer 2: yea there's a team asking for clarification. I've added a change.

This tool checks the `lastUpdateTime` of the `Progressing` block in the `status` field of the Deployment or DeploymentConfig. If it is close to a year old, but not over, a warning email will be sent. If the deployment is managed by ArgoCD then a warning email is sent asking for action to be taken. This is because ArgoCD will just scale it back up after the tool scaled it down. If not managed by ArgoCD and the timestamp is over a year then the deployment is scaled to zero replicas and an email is sent.

You can check the timestamp on your deployment in the YAML under `status`.
 
 ```yaml
   conditions:
  - lastTransitionTime: "2021-11-25T17:04:01Z"
    lastUpdateTime: "2021-11-25T17:04:11Z"
    message: replication controller "backup-postgresql-1" successfully rolled out
    reason: NewReplicationControllerAvailable
    status: "True"
    type: Progressing
  ```

> ## Deployments, deploymentconfigs and statefulsets that are crashing constantly

Reviewer: ```suggestion

## Deployments, deploymentconfigs and statefulsets that are crashing constantly
```
extra line between previous code block and next header

> This tool looks at the restart count of pods. If a pod has more than 100 restarts it looks for the controlling object of the pod, be it aDeployment, DeploymentConfig, or StatefulSet. If the object is managed by ArgoCD then a warning email is sent asking for action to be taken. This is because ArgoCD will just scale it back up after the tool scaled it down. If not managed by ArgoCD, the object is scaled to zero replicas and an email is sent.

Reviewer: ```suggestion
This tool looks at the restart count of pods. If a pod has more than 100 restarts it looks for the controlling object of the pod, be it a Deployment, DeploymentConfig, or StatefulSet. If the object is managed by ArgoCD then a warning email is sent asking for action to be taken. This is because ArgoCD will just scale it back up after the tool scaled it down. If not managed by ArgoCD, the object is scaled to zero replicas and an email is sent.
```

 
If your application hasn't been updated in over a year, your technical team should create a new image. This ensures that your application can benefit from any security patches released in the past year. It's highly probable that your image has critical security vulnerabilities if it hasn't been updated for more than a year.
 
 Maintaining your application is part of the [Memorandum of Understanding](https://digital.gov.bc.ca/cloud/services/private/onboard/#memorandum) that teams working on the platform agree to. More detailed docs on [maintaining an image](https://docs.developer.gov.bc.ca/maintain-an-application/#maintain-images) are also available.

Once you have fixed these issues, you can follow the instructions in [RedHat’s documentation on editing deployments](https://docs.openshift.com/container-platform/4.12/applications/deployments/deployment-strategies.html#odc-editing-deployments_rolling-strategy) to scale your applications back up again. 
 
Be aware that if you scale your deployment back up without addressing the underlying issues, our automated process will scale the application back down again the following week.

## Timeline 

**Silver cluster timeline**

> - Beginning January 23, 2024 these automated tools will run every Tuesday morning

Reviewer 3: ```suggestion
- Beginning January 23, 2024 these automated tools will run every Tuesday morning.
```

---
title: Platform automation

slug: platform-automation

description: Describes the automation we run on to manage the OpenShift platform.

keywords: resource, reclamation, tool, automation, security

page_purpose: Describes the automation we run on to manage the OpenShift platform.

audience: technical lead, developer

author: Matt Spencer, Steven Barre

content_owner: Olena Mitovska

sort_order: 6
---

# Platform Automation

We use some automation on the platform to help manage resource usage, alert about misconfigured objects and to encourage teams to use images that are secure and up to date. 

> To date, resource utilization (the use of CPU and RAM) on our platforms has not been efficient. For example, almost 50% of CPU capacity on the Silver cluster is reserved by apps, but not actually used. This unused utilization increases costs and impacts our ability to onboard new teams.  Sometimes teams do not maintain and update the images that their deployments are based on. This can lead to security vulnerabilities, incompatibilities and other problems  

Author: Slight change of wording here to emphasise that this is a problem relating to how applications on the platform utilize resources, not a problem with the platform itself.

 
To address this, we’ve developed automated tools that monitor apps on the platform. 
 
These tools looks for these types of objects: 
- `Deployments` and `deploymentconfigs` that are based on images have not changed in over a year

> - `Deployments`, `deploymentconfigs` and `statefulsets` that are crashing constantly 

Author: To me the suggested wording implies there are other types of objects that the tool searches for. The list aims to be complete, so I'd prefer to leave out the 'including'.

## Responding when deployments are scaled down by these tools

Before you simply scale your pods back up, you should consider fixing the underlying issues that placed your application in one of these two groups in the first place.

 If your application is crashing regularly, your technical team should investigate and fix the cause of these crashes. You can see the number of times your pod has restarted in the web console or via `oc get pods` to get an idea of if it is crashing a lot.
 
 If your application has not changed in over a year, your technical team should build a new image, which ensures that your application is able to take advantage of any security patches from the past year. It is highly likely your image contains Critical security vulnerabilities if it has not been updated in over a year.
 

>  Maintaining your application is part of the [Memorandum of Understanding](https://digital.gov.bc.ca/cloud/services/private/onboard/#memorandum) that teams working on the platform agree to. More detailed docs on [maintaining an image](https://docs.developer.gov.bc.ca/maintain-an-application/#maintain-images) are also available.

Reviewer 3: ```suggestion
Keeping your application up-to-date is a requirement outlined in the [Memorandum of Understanding](https://digital.gov.bc.ca/cloud/services/private/onboard/#memorandum) agreed upon by teams operating on the platform. Additional comprehensive documentation on [maintaining an image](https://docs.developer.gov.bc.ca/maintain-an-application/#maintain-images) is also accessible for more detailed guidance.
```

Author: I think we need to be careful with wording here. The MOU includes a line for teams responsibility to 'Build, deploy and maintain your applications'. There is no specific reference to 'Keeping your application up-to-date'

Reviewer 3: I thought about this one - you are right, the word "maintaining" is making reference to a continuous effort and it is used consistently throughout our documentation. You can leave it as such.

I used the word "keeping" as it is more of a word of everyday use, commonly known and to the point.

Once you have fixed these issues, you can follow the instructions in [RedHat’s documentation on editing deployments](https://docs.openshift.com/container-platform/4.12/applications/deployments/deployment-strategies.html#odc-editing-deployments_rolling-strategy) to scale your applications back up again. 
 
Please be aware that if you scale your deployment back up without fixing the underlying issues, our automated process will scale the application back down again the following week.  

## Timeline 

**Silver cluster timeline**
- Starting January 23, these automated tools will run weekly on Tuesday mornings

**Gold and Emerald cluster timeline** 

> - Implementation is planned for early 2024 

Author: I think in this case it is important to communicate the intending changes for Gold/Emerald. Do you have a suggestion of a different way to do this?

> - Implementation is planned for early 2024 

Author: Or perhaps we can schedule a date to update the doc?

> - Implementation is planned for early 2024 

Author: I discussed with Cailey and she made a suggestion on how to approach this, I've made a new commit with a change for this section.
